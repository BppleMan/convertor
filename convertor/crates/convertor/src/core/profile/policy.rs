use crate::core::error::ParseError;
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;
use std::collections::HashMap;
use std::str::FromStr;
use std::sync::LazyLock;

static OPTION_RANK: LazyLock<HashMap<Option<&str>, usize>> = LazyLock::new(|| {
    [None, Some("no-resolve"), Some("force-remote-dns")]
        .into_iter()
        .enumerate()
        .map(|(i, option)| (option, i))
        .collect()
});

#[derive(Default, Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
#[serde(try_from = "PolicySerialHelper")]
pub struct Policy {
    pub name: String,
    pub option: Option<String>,
    pub is_subscription: bool,
}

impl Policy {
    pub fn new(name: impl AsRef<str>, option: Option<&str>, is_subscription: bool) -> Self {
        Policy {
            name: name.as_ref().to_string(),
            option: option.map(|s| s.to_string()),
            is_subscription,
        }
    }

    pub fn subscription_policy() -> Self {
        Policy {
            name: "DIRECT".to_string(),
            option: None,
            is_subscription: true,
        }
    }

    pub fn direct_policy(option: Option<&str>) -> Self {
        Policy {
            name: "DIRECT".to_string(),
            option: option.map(|s| s.to_string()),
            is_subscription: false,
        }
    }

    pub fn is_subscription_policy(&self) -> bool {
        self.is_subscription && self.name == "DIRECT"
    }

    pub fn is_built_in(&self) -> bool {
        (self.name == "DIRECT" || self.name == "REJECT" || self.name == "FINAL") && !self.is_subscription
    }
}

impl FromStr for Policy {
    type Err = ParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let parts = s.splitn(2, ',').map(str::trim).collect::<Vec<_>>();
        if parts.is_empty() {
            return Err(ParseError::Policy {
                line: 0,
                reason: format!("无法理解的策略\"{}\"", s),
            });
        }
        Ok(Policy {
            name: parts[0].to_string(),
            option: parts.get(1).map(|part| part.to_string()),
            is_subscription: false,
        })
    }
}

impl PartialOrd<Policy> for Policy {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Policy {
    fn cmp(&self, other: &Self) -> Ordering {
        let option_rank = |option: &Option<String>| {
            *OPTION_RANK
                .get(&option.as_ref().map(String::as_str))
                .unwrap_or(&usize::MAX)
        };

        self.is_subscription
            .cmp(&other.is_subscription)
            .reverse()
            .then(self.name.cmp(&other.name))
            .then(option_rank(&self.option).cmp(&option_rank(&other.option)))
    }
}

#[derive(Deserialize)]
#[serde(untagged)]
enum PolicySerialHelper {
    Str(String),
    Obj {
        name: String,
        option: Option<String>,
        #[serde(default)]
        is_subscription: bool,
    },
}

impl TryFrom<PolicySerialHelper> for Policy {
    type Error = ParseError;

    fn try_from(repr: PolicySerialHelper) -> Result<Self, Self::Error> {
        match repr {
            PolicySerialHelper::Str(s) => Policy::from_str(&s),
            PolicySerialHelper::Obj {
                name,
                option,
                is_subscription,
            } => {
                if name.trim().is_empty() {
                    return Err(ParseError::Policy {
                        line: 0,
                        reason: "策略名称不能为空".to_string(),
                    });
                }
                Ok(Policy {
                    name,
                    option,
                    is_subscription,
                })
            }
        }
    }
}

// impl<'de> Deserialize<'de> for Policy {
//     fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
//     where
//         D: Deserializer<'de>,
//     {
//         struct PolicyVisitor;
//
//         impl<'de> serde::de::Visitor<'de> for PolicyVisitor {
//             type Value = Policy;
//
//             fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
//                 write!(formatter, "策略语法应该形如: 策略名称[,选项]")
//             }
//
//             fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
//             where
//                 E: serde::de::Error,
//             {
//                 Policy::from_str(v).map_err(E::custom)
//             }
//         }
//
//         #[derive(Deserialize)]
//         pub struct SerializePolicy {
//             pub name: String,
//             pub option: Option<String>,
//             pub is_subscription: bool,
//         }
//
//         match deserializer.deserialize_str(PolicyVisitor) {
//             Ok(policy) => Ok(policy),
//             Err(err) => {
//                 let sp = SerializePolicy::deserialize(deserializer)?;
//             }
//         }
//     }
// }
