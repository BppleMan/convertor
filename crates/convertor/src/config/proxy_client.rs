use clap::ValueEnum;
use serde::{Deserialize, Serialize};
use std::fmt::{Display, Formatter};
use std::str::FromStr;

#[derive(Default, Debug, Copy, Clone, Eq, PartialEq, Hash)]
#[derive(ValueEnum, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ProxyClient {
    #[default]
    Surge,
    Clash,
}

impl ProxyClient {
    pub fn as_str(&self) -> &'static str {
        match self {
            ProxyClient::Surge => "surge",
            ProxyClient::Clash => "clash",
        }
    }

    pub fn variants() -> &'static [Self] {
        &[Self::Surge, Self::Clash]
    }
}

impl Display for ProxyClient {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

impl FromStr for ProxyClient {
    type Err = String;

    fn from_str(s: &str) -> core::result::Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "surge" => Ok(ProxyClient::Surge),
            "clash" => Ok(ProxyClient::Clash),
            _ => Err(format!("Invalid proxy client: {}", s)),
        }
    }
}
