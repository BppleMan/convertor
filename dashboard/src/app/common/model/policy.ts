export class Policy {
    public constructor(
        public name: string,
        public option: string | null,
        public is_subscription: boolean,
    ) {
    }

    public static deserialize(policy: Policy) {
        return new Policy(
            policy.name,
            policy.option,
            policy.is_subscription,
        );
    }
}
