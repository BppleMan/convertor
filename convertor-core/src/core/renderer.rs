#![deny(unused, unused_variables)]

use crate::client::Client;
use crate::core::profile::policy::Policy;
use crate::core::profile::profile::Profile;
use crate::core::profile::proxy::Proxy;
use crate::core::profile::proxy_group::ProxyGroup;
use crate::core::profile::rule::{ProviderRule, Rule};
use crate::core::profile::rule_provider::RuleProvider;
use crate::core::result::RenderResult;
use std::fmt::Write;
use tracing::instrument;

pub mod surge_renderer;
pub mod clash_renderer;

pub const INDENT: usize = 4;

pub trait Renderer {
    type PROFILE: Profile;

    fn client() -> Client;

    fn render_profile(profile: &Self::PROFILE) -> RenderResult<String>;

    fn render_general(profile: &Self::PROFILE) -> RenderResult<String>;

    #[instrument(skip_all)]
    fn render_proxies(proxies: &[Proxy]) -> RenderResult<String> {
        Self::render_lines(proxies, Self::render_proxy)
    }

    fn render_proxy(proxy: &Proxy) -> RenderResult<String>;

    #[instrument(skip_all)]
    fn render_proxy_groups(proxy_groups: &[ProxyGroup]) -> RenderResult<String> {
        Self::render_lines(proxy_groups, Self::render_proxy_group)
    }

    fn render_proxy_group(proxy_group: &ProxyGroup) -> RenderResult<String>;

    #[instrument(skip_all)]
    fn render_rules(rules: &[Rule]) -> RenderResult<String> {
        Self::render_lines(rules, Self::render_rule)
    }

    fn render_rule(rule: &Rule) -> RenderResult<String>;

    /// 适用于渲染规则集类型的规则，不包含注释
    fn render_rule_for_provider(rule: &Rule) -> RenderResult<String>;

    #[instrument(skip_all)]
    fn render_provider_rules(rules: &[ProviderRule]) -> RenderResult<String> {
        let mut output = String::new();
        match Self::client() {
            Client::Surge => {
                writeln!(output, "{}", Self::render_lines(rules, Self::render_provider_rule)?)?;
            }
            Client::Clash => {
                writeln!(output, "payload:")?;
                writeln!(output, "{}", Self::render_lines(rules, Self::render_provider_rule)?)?;
            }
        }
        Ok(output)
    }

    fn render_provider_rule(rule: &ProviderRule) -> RenderResult<String>;

    fn render_policy(policy: &Policy) -> RenderResult<String> {
        let mut output = String::new();
        write!(output, "{}", policy.name)?;
        if let Some(option) = &policy.option {
            write!(output, ",{}", option)?;
        }
        Ok(output)
    }

    fn render_rule_providers(rule_providers: &[(String, RuleProvider)]) -> RenderResult<String>;

    fn render_rule_provider(rule_provider: &(String, RuleProvider)) -> RenderResult<String>;

    fn render_provider_name_for_policy(policy: &Policy) -> RenderResult<String>;

    fn render_lines<T, F>(lines: impl IntoIterator<Item = T>, map: F) -> RenderResult<String>
    where
        F: FnMut(T) -> RenderResult<String>,
    {
        let output = lines
            .into_iter()
            .map(map)
            .map(|line| match Self::client() {
                Client::Surge => line,
                Client::Clash => line.map(Self::indent_line),
            })
            .collect::<RenderResult<Vec<_>>>()?
            .join("\n");
        Ok(output)
    }

    fn indent_line(line: String) -> String {
        format!("{:indent$}{}", "", format_args!("- {}", line), indent = INDENT)
    }
}
