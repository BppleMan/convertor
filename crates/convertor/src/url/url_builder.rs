use crate::common::encrypt::encrypt;
use crate::config::client_config::ProxyClient;
use crate::config::provider_config::Provider;
use crate::core::profile::policy::Policy;
use crate::core::profile::surge_header::SurgeHeader;
use crate::error::UrlBuilderError;
use crate::url::convertor_url::{ConvertorUrl, ConvertorUrlType};
use crate::url::query::ConvertorQuery;
use serde::{Deserialize, Serialize};
use url::Url;

#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
pub struct UrlBuilder {
    pub secret: String,
    pub enc_secret: String,
    pub client: ProxyClient,
    pub provider: Provider,
    pub server: Url,
    pub sub_url: Url,
    pub enc_sub_url: String,
    pub interval: u64,
    pub strict: bool,
}

impl UrlBuilder {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        secret: impl AsRef<str>,
        enc_secret: Option<String>,
        client: ProxyClient,
        provider: Provider,
        server: Url,
        sub_url: Url,
        enc_sub_url: Option<String>,
        interval: u64,
        strict: bool,
    ) -> Result<Self, UrlBuilderError> {
        let secret = secret.as_ref().to_string();
        let enc_secret = enc_secret
            .map(Ok)
            .unwrap_or_else(|| encrypt(secret.as_bytes(), secret.as_str()))?;
        let enc_sub_url = enc_sub_url
            .map(Ok)
            .unwrap_or_else(|| encrypt(secret.as_bytes(), sub_url.as_str()))?;

        let builder = Self {
            secret,
            enc_secret,
            client,
            provider,
            server,
            sub_url,
            enc_sub_url,
            interval,
            strict,
        };
        Ok(builder)
    }

    pub fn from_convertor_query(
        query: ConvertorQuery,
        secret: impl AsRef<str>,
        client: ProxyClient,
        provider: Provider,
    ) -> Result<Self, UrlBuilderError> {
        let ConvertorQuery {
            server,
            sub_url,
            enc_sub_url,
            interval,
            strict,
            secret: secret_opt,
            enc_secret,
            policy: _,
        } = query;
        let secret = secret_opt.unwrap_or(secret.as_ref().to_string());
        let strict = strict.unwrap_or(true);
        Self::new(
            secret,
            enc_secret,
            client,
            provider,
            server,
            sub_url,
            Some(enc_sub_url),
            interval,
            strict,
        )
    }

    pub fn build_raw_url(&self) -> ConvertorUrl {
        let mut url = self.sub_url.clone();
        url.query_pairs_mut().append_pair("flag", self.client.as_ref());
        ConvertorUrl::raw(url)
    }

    pub fn build_raw_profile_url(&self) -> Result<ConvertorUrl, UrlBuilderError> {
        let query = self.as_profile_query().encode_to_profile_query()?;
        let url = ConvertorUrl::raw_profile(
            self.server.clone(),
            format!(
                "/raw-profile/{}/{}",
                self.client.as_ref().to_ascii_lowercase(),
                self.provider.as_ref().to_ascii_lowercase()
            ),
            query,
        );
        Ok(url)
    }

    pub fn build_profile_url(&self) -> Result<ConvertorUrl, UrlBuilderError> {
        let query = self.as_profile_query().encode_to_profile_query()?;
        let url = ConvertorUrl::profile(
            self.server.clone(),
            format!(
                "/profile/{}/{}",
                self.client.as_ref().to_ascii_lowercase(),
                self.provider.as_ref().to_ascii_lowercase()
            ),
            query,
        );
        Ok(url)
    }

    pub fn build_rule_provider_url(&self, policy: &Policy) -> Result<ConvertorUrl, UrlBuilderError> {
        let query = self.as_rule_provider_query(policy).encode_to_rule_provider_query()?;
        let url = ConvertorUrl::rule_provider(
            policy.clone(),
            self.server.clone(),
            format!(
                "/rule-provider/{}/{}",
                self.client.as_ref().to_ascii_lowercase(),
                self.provider.as_ref().to_ascii_lowercase()
            ),
            query,
        );
        Ok(url)
    }

    pub fn build_sub_logs_url(&self) -> Result<ConvertorUrl, UrlBuilderError> {
        let query = self.as_sub_logs_query().encode_to_sub_logs_query()?;
        let url = ConvertorUrl::sub_logs(self.server.clone(), "/api/sub-logs", query);
        Ok(url)
    }

    // 构造专属 Surge 的订阅头
    pub fn build_surge_header(&self, r#type: ConvertorUrlType) -> Result<SurgeHeader, UrlBuilderError> {
        let url = match r#type {
            ConvertorUrlType::Raw => self.build_raw_url(),
            ConvertorUrlType::RawProfile => self.build_raw_profile_url()?,
            ConvertorUrlType::Profile => self.build_profile_url()?,
            _ => return Err(UrlBuilderError::UnsupportedUrlType(r#type)),
        };
        Ok(SurgeHeader::new(url, self.interval, self.strict))
    }
}

impl UrlBuilder {
    pub fn as_profile_query(&self) -> ConvertorQuery {
        ConvertorQuery {
            server: self.server.clone(),
            sub_url: self.sub_url.clone(),
            enc_sub_url: self.enc_sub_url.clone(),
            interval: self.interval,
            strict: Some(self.strict),
            policy: None,
            secret: None,
            enc_secret: None,
        }
    }

    pub fn as_rule_provider_query(&self, policy: &Policy) -> ConvertorQuery {
        let mut query = self.as_profile_query();
        query.policy = Some(policy.clone());
        query
    }

    pub fn as_sub_logs_query(&self) -> ConvertorQuery {
        let mut query = self.as_profile_query();
        query.secret = Some(self.secret.clone());
        query.enc_secret = Some(self.enc_secret.clone());
        query
    }
}

pub trait HostPort {
    fn host_port(&self) -> Option<String>;
}

impl HostPort for Url {
    fn host_port(&self) -> Option<String> {
        match (self.host_str(), self.port()) {
            (Some(host), Some(port)) => Some(format!("{host}:{port}")),
            (Some(host), None) => Some(host.to_string()),
            _ => None,
        }
    }
}
