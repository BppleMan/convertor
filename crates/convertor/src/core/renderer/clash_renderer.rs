use crate::config::client_config::ProxyClient;
use crate::core::profile::clash_profile::ClashProfile;
use crate::core::profile::policy::Policy;
use crate::core::profile::proxy::Proxy;
use crate::core::profile::proxy_group::ProxyGroup;
use crate::core::profile::rule::{ProviderRule, Rule};
use crate::core::profile::rule_provider::RuleProvider;
use crate::core::renderer::{INDENT, Renderer};
use crate::error::RenderError;
use std::fmt::Write;
use tracing::instrument;

type Result<T> = core::result::Result<T, RenderError>;

pub struct ClashRenderer;

impl Renderer for ClashRenderer {
    type PROFILE = ClashProfile;

    fn client() -> ProxyClient {
        ProxyClient::Clash
    }

    #[instrument(skip_all)]
    fn render_profile(profile: &Self::PROFILE) -> Result<String> {
        let mut output = String::new();
        writeln!(output, "{}", Self::render_general(profile)?)?;

        let proxies = Self::render_proxies(&profile.proxies)?;
        writeln!(output, "proxies:")?;
        writeln!(output, "{proxies}")?;

        let proxy_groups = Self::render_proxy_groups(&profile.proxy_groups)?;
        writeln!(output, "proxy-groups:")?;
        writeln!(output, "{proxy_groups}")?;

        let rule_providers = Self::render_rule_providers(&profile.rule_providers)?;
        writeln!(output, "rule-providers:")?;
        writeln!(output, "{rule_providers}")?;

        let rules = Self::render_rules(&profile.rules)?;
        writeln!(output, "rules:")?;
        writeln!(output, "{rules}")?;

        Ok(output)
    }

    #[instrument(skip_all)]
    fn render_general(profile: &Self::PROFILE) -> Result<String> {
        let mut output = String::new();
        writeln!(output, "port: {}", profile.port)?;
        writeln!(output, "socks-port: {}", profile.socks_port)?;
        writeln!(output, "redir-port: {}", profile.redir_port)?;
        writeln!(output, "allow-lan: {}", profile.allow_lan)?;
        writeln!(output, "mode: {}", profile.mode)?;
        writeln!(output, "log-level: {}", profile.log_level)?;
        writeln!(output, r#"external-controller: {}"#, profile.external_controller)?;
        writeln!(output, r#"external-ui: {}"#, profile.external_ui)?;
        if let Some(secret) = &profile.secret {
            writeln!(output, r#"secret: "{secret}""#)?;
        }
        Ok(output)
    }

    fn render_proxy(proxy: &Proxy) -> Result<String> {
        let mut output = String::new();
        write!(output, "{{ ")?;
        write!(output, r#"name: "{}""#, &proxy.name)?;
        write!(output, r#", type: "{}""#, &proxy.r#type)?;
        write!(output, r#", server: "{}""#, &proxy.server)?;
        write!(output, r#", port: {}"#, &proxy.port)?;
        write!(output, r#", password: "{}""#, &proxy.password)?;
        if let Some(udp) = &proxy.udp {
            write!(output, r#", udp: {udp}"#)?;
        }
        if let Some(tfo) = &proxy.tfo {
            write!(output, r#", tfo: {tfo}"#)?;
        }
        if let Some(cipher) = &proxy.cipher {
            write!(output, r#", cipher: {cipher}"#)?;
        }
        if let Some(sni) = &proxy.sni {
            write!(output, r#", sni: "{sni}""#)?;
        }
        if let Some(skip_cert_verify) = &proxy.skip_cert_verify {
            write!(output, r#", skip-cert-verify: {skip_cert_verify}"#)?;
        }
        write!(output, " }}")?;
        Ok(output)
    }

    fn render_proxy_group(proxy_group: &ProxyGroup) -> Result<String> {
        let mut output = String::new();
        write!(output, "{{ ")?;
        write!(output, r#"name: "{}""#, proxy_group.name)?;
        write!(output, r#", type: "{}""#, proxy_group.r#type.as_str())?;
        write!(output, r#", proxies: [ {} ]"#, proxy_group.proxies.join(", "))?;
        write!(output, " }}")?;
        Ok(output)
    }

    fn render_rule(rule: &Rule) -> Result<String> {
        let mut output = String::new();
        write!(output, "{}", rule.rule_type.as_str())?;
        if let Some(value) = &rule.value {
            write!(output, ",{value}")?;
        }
        write!(output, ",{}", Self::render_policy(&rule.policy)?)?;
        Ok(output)
    }

    fn render_rule_for_provider(rule: &Rule) -> Result<String> {
        Self::render_rule(rule)
    }

    fn render_provider_rule(rule: &ProviderRule) -> Result<String> {
        Ok(format!("{},{}", rule.rule_type.as_str(), rule.value))
    }

    #[instrument(skip_all)]
    fn render_rule_providers(rule_providers: &[(String, RuleProvider)]) -> Result<String> {
        let output = rule_providers
            .iter()
            .map(Self::render_rule_provider)
            .map(|line| line.map(|line| format!("{:indent$}{}", "", line, indent = INDENT)))
            .collect::<Result<Vec<_>>>()?
            .join("\n");
        Ok(output)
    }

    fn render_rule_provider(rule_provider: &(String, RuleProvider)) -> Result<String> {
        let (name, rule_provider) = rule_provider;
        Ok(format!(
            r#"{}: {{ type: "{}", url: "{}", path: "{}", interval: {}, size-limit: {}, format: "{}", behavior: "{}" }}"#,
            name,
            rule_provider.r#type,
            rule_provider.url,
            rule_provider.path,
            rule_provider.interval,
            rule_provider.size_limit,
            rule_provider.format,
            rule_provider.behavior
        ))
    }

    fn render_provider_name_for_policy(policy: &Policy) -> String {
        let mut output = if policy.is_subscription {
            "Subscription".to_string()
        } else {
            policy.name.clone()
        };
        match &policy.option {
            Some(option) => {
                output.push('_');
                output.push_str(option.replace('-', "_").as_str());
            }
            None => {
                output.push_str("_policy");
            }
        }
        output
    }
}
