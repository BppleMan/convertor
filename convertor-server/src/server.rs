use axum::Router;
use axum::extract::{RawQuery, State};
use axum::http::StatusCode;
use axum::http::header::ToStrError;
use axum::response::{IntoResponse, Response};
use axum::routing::get;
use color_eyre::Result;
use color_eyre::eyre::{WrapErr, eyre};
use convertor_core::api::ServiceApi;
use convertor_core::client::Client;
use convertor_core::config::ConvertorConfig;
use convertor_core::core::profile::clash_profile::ClashProfile;
use convertor_core::core::profile::surge_profile::SurgeProfile;
use convertor_core::url::{ConvertorUrl, ConvertorUrlError};
use moka::future::Cache;
use std::net::{SocketAddr, SocketAddrV4};
use std::path::Path;
use std::sync::Arc;
use surge_router::sub_logs;
use thiserror::Error;
use tokio::signal;
use tower_http::LatencyUnit;
use tower_http::trace::{DefaultMakeSpan, DefaultOnRequest, DefaultOnResponse, TraceLayer};
use tracing::instrument;
use tracing::{info, warn};

pub mod clash_router;
pub mod surge_router;

pub struct AppState {
    pub config: ConvertorConfig,
    pub api: ServiceApi,
    pub profile_cache: Cache<ConvertorUrl, String>,
    pub surge_cache: Cache<ConvertorUrl, SurgeProfile>,
    pub clash_cache: Cache<ConvertorUrl, ClashProfile>,
}

impl AppState {
    pub fn new(config: ConvertorConfig, api: ServiceApi) -> Self {
        let duration = std::time::Duration::from_secs(60 * 60); // 1 hour
        let profile_cache = Cache::builder().max_capacity(100).time_to_live(duration).build();
        let surge_cache = Cache::builder().max_capacity(100).time_to_live(duration).build();
        let clash_cache = Cache::builder().max_capacity(100).time_to_live(duration).build();
        Self {
            config,
            api,
            profile_cache,
            surge_cache,
            clash_cache,
        }
    }
}

#[derive(Debug, Error)]
pub enum AppError {
    #[error(transparent)]
    ConvertorUrl(#[from] ConvertorUrlError),

    #[error("Unauthorized: {0}")]
    Unauthorized(String),

    #[error(transparent)]
    Eyre(#[from] color_eyre::eyre::Error),

    #[error(transparent)]
    ToStr(#[from] ToStrError),

    #[error(transparent)]
    Utf8Error(#[from] std::str::Utf8Error),

    #[error(transparent)]
    CacheError(#[from] Arc<AppError>),
}

impl IntoResponse for AppError {
    fn into_response(self) -> Response {
        let message = format!("{self:?}");
        let message = console::strip_ansi_codes(&message).to_string();
        (StatusCode::INTERNAL_SERVER_ERROR, message).into_response()
    }
}

pub async fn start_server(
    listen_addr: SocketAddrV4,
    config: ConvertorConfig,
    api: ServiceApi,
    base_dir: impl AsRef<Path>,
) -> Result<()> {
    info!("base_dir: {}", base_dir.as_ref().display());
    info!("监听中: {}", &listen_addr);
    warn!("建议使用 nginx 等网关进行反向代理，以开启 HTTPS 支持");
    info!("服务启动，使用 Ctrl+C 或 SIGTERM 关闭服务");
    let listener = tokio::net::TcpListener::bind(listen_addr).await?;

    let app_state = AppState::new(config, api);
    let app = router(app_state);
    axum::serve(listener, app.into_make_service_with_connect_info::<SocketAddr>())
        .with_graceful_shutdown(shutdown_signal())
        .await?;
    info!("服务关闭");
    Ok(())
}

pub fn router(app_state: AppState) -> Router {
    Router::new()
        .route("/profile", get(profile))
        .route("/rule-provider", get(rule_provider))
        .route("/sub-logs", get(sub_logs))
        .with_state(Arc::new(app_state))
        .layer(
            TraceLayer::new_for_http()
                .make_span_with(DefaultMakeSpan::new().include_headers(true))
                .on_request(DefaultOnRequest::new().level(tracing::Level::INFO))
                .on_response(
                    DefaultOnResponse::new()
                        .level(tracing::Level::INFO)
                        .latency_unit(LatencyUnit::Millis),
                ),
        )
}

#[instrument(skip_all)]
pub async fn profile(State(state): State<Arc<AppState>>, RawQuery(query): RawQuery) -> Result<String, AppError> {
    let convertor_url = query
        .map(|query| ConvertorUrl::parse_from_query_string(query, &state.config.secret))
        .ok_or_else(|| eyre!("查询参数不能为空"))?
        .wrap_err("解析查询字符串失败")?;
    let client = convertor_url.client;
    let profile = state
        .clone()
        .profile_cache
        .try_get_with(convertor_url.clone(), async {
            let raw_profile = state.api.get_raw_profile(client).await?;
            let profile = match client {
                Client::Surge => surge_router::profile_impl(state, convertor_url, raw_profile).await,
                Client::Clash => clash_router::profile_impl(state, convertor_url, raw_profile).await,
            }?;
            Ok::<_, AppError>(profile)
        })
        .await?;
    Ok(profile)
}

#[instrument(skip_all)]
pub async fn rule_provider(State(state): State<Arc<AppState>>, RawQuery(query): RawQuery) -> Result<String, AppError> {
    let convertor_url = query
        .map(|query| ConvertorUrl::parse_from_query_string(query, &state.config.secret))
        .ok_or_else(|| eyre!("查询参数不能为空"))?
        .wrap_err("解析查询字符串失败")?;
    let client = &convertor_url.client;
    let policy = convertor_url.policy.clone();
    let raw_profile = state.api.get_raw_profile(*client).await?;
    match (client, policy) {
        (Client::Surge, Some(policy)) => {
            surge_router::rule_provider_impl(state, convertor_url, raw_profile, policy.into()).await
        }
        (Client::Clash, Some(policy)) => {
            clash_router::rule_provider_impl(state, convertor_url, raw_profile, policy.into()).await
        }
        _ => Err(eyre!("错误的 client 或 policy 参数")),
    }
    .map_err(Into::into)
}

pub async fn shutdown_signal() {
    let ctrl_c = async {
        signal::ctrl_c().await.expect("failed to install Ctrl+C handler");
    };

    #[cfg(unix)]
    let terminate = async {
        signal::unix::signal(signal::unix::SignalKind::terminate())
            .expect("failed to install signal handler")
            .recv()
            .await;
    };

    #[cfg(not(unix))]
    let terminate = std::future::pending::<()>();

    tokio::select! {
        _ = ctrl_c => {},
        _ = terminate => {},
    }
}
