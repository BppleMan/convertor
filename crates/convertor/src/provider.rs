use crate::common::cache::Cache;
use crate::config::subscription_config::Headers;
use crate::error::{ApiFailed, ProviderError};
use redis::aio::ConnectionManager;
use reqwest::Method;
use std::ops::Deref;
use std::time::Duration;
use url::Url;

#[derive(Clone)]
pub struct SubscriptionProvider {
    pub client: reqwest::Client,
    pub cache: Cache<String, String>,
}

impl SubscriptionProvider {
    pub fn new(redis: Option<ConnectionManager>) -> Self {
        let client = reqwest::Client::new();
        let cache = Cache::new(
            redis,
            10,
            Duration::from_secs(60 * 60),
            Duration::from_secs(60 * 60 * 12),
        );
        Self { client, cache }
    }

    pub async fn get_raw_profile(&self, sub_url: Url, headers: Headers) -> Result<String, ProviderError> {
        let mut request_builder = self.client.request(Method::GET, sub_url.clone());
        for (k, v) in headers.deref() {
            request_builder = request_builder.header(k, v);
        }
        let request = request_builder.build().map_err(|e| ProviderError::RequestError {
            reason: "无法构建请求".to_string(),
            source: Box::new(e),
        })?;
        let response = self
            .client
            .execute(request)
            .await
            .map_err(|e| ProviderError::RequestError {
                reason: "请求失败".to_string(),
                source: Box::new(e),
            })?;
        let response_status = response.status();
        let response_headers = response
            .headers()
            .into_iter()
            .flat_map(|(k, v)| {
                let k = k.to_string();
                v.to_str().map(|v| (k, v.to_string())).ok()
            })
            .collect::<Headers>();
        let response_body = response.text().await.map_err(|e| ProviderError::RequestError {
            reason: "读取响应体失败".to_string(),
            source: Box::new(e),
        })?;
        if response_status.is_success() {
            Ok(response_body)
        } else {
            Err(ProviderError::ApiFailed(Box::new(ApiFailed {
                request_url: sub_url,
                request_method: Method::GET,
                request_headers: headers,
                response_status,
                response_headers,
                response_body,
            })))
        }
    }
}
