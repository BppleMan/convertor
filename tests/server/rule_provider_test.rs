use crate::init_test;
use crate::server::{ServerContext, start_server};
use axum::body::Body;
use axum::extract::Request;
use convertor::common::config::provider::SubProvider;
use convertor::common::config::proxy_client::ProxyClient;
use convertor::core::profile::policy::Policy;
use convertor::core::renderer::Renderer;
use convertor::core::renderer::clash_renderer::ClashRenderer;
use convertor::core::url_builder::HostPort;
use http_body_util::BodyExt;
use tower::ServiceExt;

async fn rule_provider(
    server_context: &ServerContext,
    client: ProxyClient,
    provider: SubProvider,
    policy: Policy,
) -> color_eyre::Result<String> {
    let ServerContext { app, app_state, .. } = server_context;
    let url_builder = app_state.config.create_url_builder(client, provider)?;

    let rule_provider_url = url_builder.build_rule_provider_url(&policy)?;
    let request = Request::builder()
        .uri(rule_provider_url.to_string())
        .method("GET")
        .header("host", "127.0.0.1")
        .header("user-agent", concat!("convertor/", env!("CARGO_PKG_VERSION")))
        .body(Body::empty())?;
    let response = app.clone().oneshot(request).await?;

    let actual = String::from_utf8_lossy(&response.into_body().collect().await?.to_bytes())
        .to_string()
        .replace(&url_builder.sub_url.host_port()?, "localhost");

    Ok(actual)
}

fn policies() -> [Policy; 7] {
    [
        Policy::subscription_policy(),
        Policy::new("BosLife", None, false),
        Policy::new("BosLife", Some("no-resolve"), false),
        Policy::new("BosLife", Some("force-remote-dns"), false),
        Policy::direct_policy(None),
        Policy::direct_policy(Some("no-resolve")),
        Policy::direct_policy(Some("force-remote-dns")),
    ]
}

#[tokio::test]
async fn test_rule_provider_surge_boslife() -> color_eyre::Result<()> {
    init_test();
    let server_context = start_server().await?;
    let policies = policies();
    for policy in policies {
        let ctx = format!(
            "test_rule_provider_surge_boslife_{}",
            ClashRenderer::render_provider_name_for_policy(&policy)?
        );
        let actual = rule_provider(&server_context, ProxyClient::Surge, SubProvider::BosLife, policy).await?;
        insta::assert_snapshot!(ctx, actual);
    }
    Ok(())
}

#[tokio::test]
async fn test_rule_provider_clash_boslife() -> color_eyre::Result<()> {
    init_test();
    let server_context = start_server().await?;
    let policies = policies();
    for policy in policies {
        let ctx = format!(
            "test_rule_provider_clash_boslife_{}",
            ClashRenderer::render_provider_name_for_policy(&policy)?
        );
        let actual = rule_provider(&server_context, ProxyClient::Clash, SubProvider::BosLife, policy).await?;
        insta::assert_snapshot!(ctx, actual);
    }
    Ok(())
}
