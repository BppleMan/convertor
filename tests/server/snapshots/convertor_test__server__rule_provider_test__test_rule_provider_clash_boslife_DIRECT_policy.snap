---
source: tests/server/rule_provider_test.rs
expression: actual
---
payload:
    - DOMAIN,trdy.subiway.com:8888
